src/Datadog.Trace/Logging/CustomSerilogLogProvider.cs

```diff
@@ -5,44 +5,80 @@
 
 using System;
 using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
 using Datadog.Trace.Logging.LogProviders;
 
 namespace Datadog.Trace.Logging
 {
     internal class CustomSerilogLogProvider : SerilogLogProvider, ILogProviderWithEnricher
     {
+        private static ConditionalWeakTable<object, object> _enricherToArrayConditionalWeakTable;
         private static Func<object, IDisposable> _pushMethod;
+        private static Func<object, IDisposable> _openContext;
 
         public CustomSerilogLogProvider()
         {
-            _pushMethod = GetPush();
-        }
+            _enricherToArrayConditionalWeakTable = new();
 
-        public IDisposable OpenContext(object enricher)
-        {
-            return _pushMethod(enricher);
+            var logEnricherType = GetLogEnricherType();
+            if (GetPushMethodInfo() != null)
+            {
+                _openContext = _pushMethod = GeneratePushDelegate(GetPushMethodInfo(), logEnricherType);
+            }
+            else
+            {
+                _pushMethod = GeneratePushDelegate(GetPushPropertiesMethodInfo(), logEnricherType.MakeArrayType());
+                _openContext = (enricher) =>
+                {
+                    // Use a conditional weak table to cache a map between the enricher and an array of size 1 containing the enricher,
+                    // since the API requires an argument of type ILogEventEnricher[]
+                    object properties = _enricherToArrayConditionalWeakTable.GetValue(key: enricher, createValueCallback: ConditionalWeakTableCallback);
```

Comment:
I think we can remove this whole ConditionalWeakTable by adding some logic in `SerilogEnricher` that knows whether to wrap the value in an array or not

Reply:
Done in 35efec447b758d7fd8fd7988af26a1447c2c8756 , let me know what you think